scripts/zones/Rabao/npcs/Irmilant.lua

```diff
@@ -3,73 +3,73 @@
+    ImmortalLuShang = player:getQuestStatus(OUTLANDS,THE_IMMORTAL_LU_SHANG)
 
-    if ((ImmortalLuShang == QUEST_ACCEPTED or ImmortalLuShang == QUEST_COMPLETED) and trade:hasItemQty(720,1) and trade:hasItemQty(489,1) and trade:hasItemQty(4102,2) and trade:getItemCount() == 4) then
-        player:startEvent(78);
-    elseif ((Indomitable == QUEST_ACCEPTED or Indomitable == QUEST_COMPLETED) and trade:hasItemQty(1837,1) and trade:hasItemQty(1826,1) and trade:getItemCount() == 2) then
-        player:startEvent(132);
+    if (ImmortalLuShang == QUEST_ACCEPTED or ImmortalLuShang == QUEST_COMPLETED) and npcUtil.tradeHas(trade, {720, 489, 4102}) then
+        player:startEvent(78)
+    elseif (Indomitable == QUEST_ACCEPTED or Indomitable == QUEST_COMPLETED) and npcUtil.tradeHas(trade, {1837, 1826}) then
+        player:startEvent(132)
     end
-end;
+end
 
 function onTrigger(player,npc)
-    Indomitable = player:getQuestStatus(OUTLANDS,INDOMITABLE_SPIRIT);
-    ImmortalLuShang = player:getQuestStatus(OUTLANDS,THE_IMMORTAL_LU_SHANG);
+    Indomitable = player:getQuestStatus(OUTLANDS,INDOMITABLE_SPIRIT)
+    ImmortalLuShang = player:getQuestStatus(OUTLANDS,THE_IMMORTAL_LU_SHANG)
     realday = tonumber(os.date("%j")); -- %M for next minute, %j for next day
 
     if (player:hasItem(489) == true and (ImmortalLuShang == QUEST_AVAILABLE or ImmortalLuShang == QUEST_COMPLETED)) then
-        player:startEvent(77); --Offer the quest if the player has the broken rod
+        player:startEvent(77) --Offer the quest if the player has the broken rod
     elseif (player:hasKeyItem(1977) == true and (Indomitable == QUEST_AVAILABLE or Indomitable == QUEST_COMPLETED)) then
-        player:startEvent(131); --Begins Indomitable Spirit
+        player:startEvent(131) --Begins Indomitable Spirit
     elseif ((Indomitable == QUEST_ACCEPTED or Indomitable == QUEST_COMPLETED) and realday == player:getVar("IndomitableSpiritVar")) then
-        player:startEvent(133); --Asks the player to wait
+        player:startEvent(133) --Asks the player to wait
     elseif ((Indomitable == QUEST_ACCEPTED or Indomitable == QUEST_COMPLETED) and realday ~= player:getVar("IndomitableSpiritVar")) then
-        player:startEvent(134); --Ends the Quest
+        player:startEvent(134) --Ends the Quest
     elseif (Indomitable == QUEST_COMPLETED) then
-        player:startEvent(135); --Dialogue for those who have completed Indomitable Spirit
+        player:startEvent(135) --Dialogue for those who have completed Indomitable Spirit
     else
-        player:startEvent(76); --Standard Dialogue
+        player:startEvent(76) --Standard Dialogue
     end
-end;
+end
 
 function onEventUpdate(player,csid,option)
-end;
+end
 
 function onEventFinish(player,csid,option)
     if (csid == 77) then
-        player:addQuest(OUTLANDS,THE_IMMORTAL_LU_SHANG);
+        player:addQuest(OUTLANDS,THE_IMMORTAL_LU_SHANG)
     elseif (csid == 78) then
```

Comment:
This whole block can be reduced to this:

```
elseif csid == 78 and npcUtil.completeQuest(player, OUTLANDS, THE_IMMORTAL_LU_SHANG, {item=17386, fameArea=RABAO, fame=60, title=dsp.title.THE_IMMORTAL_FISHER_LU_SHANG}) then
    player:confirmTrade()
elseif csid == 131 then
.
.
.
```

Note that when you use npcUtil.tradeHas(), you should typically close the trade with player:confirmTrade rather than player:tradeComplete.  tradeComplete eats all items in the trade window, whereas confirmTrade only eats ones specified in npcUtil.tradeHas. Extras are returned to player.